File: packages/orchestrator/src/node-runner.ts
/**
 * Node Runner — spawns a node as a child process, manages NDJSON stdin/stdout,
 * handles lifecycle events and clean shutdown.
 *
 * Each node is a child process that:
 * - Receives NDJSON events on stdin
 * - Emits NDJSON events on stdout
 * - Logs to stderr (forwarded to parent's stderr)
 * - Gets settings via ACPFX_SETTINGS env var
 * - Must emit lifecycle.ready when initialized
 * - Exits on stdin EOF or SIGTERM
 */

import { ChildProcess, fork, spawn as cpSpawn } from "node:child_process";
import { createInterface } from "node:readline";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { existsSync } from "node:fs";
import { type AnyEvent, parseEvent, serializeEvent } from "@acpfx/core";

const __dirname = dirname(fileURLToPath(import.meta.url));

export type NodeRunnerOptions = {
  name: string;
  use: string;
  settings?: Record<string, unknown>;
  env?: Record<string, string>;
  quiet?: boolean; // suppress stderr forwarding (when UI is active)
  onEvent: (event: AnyEvent) => void;
  onError: (error: Error) => void;
  onExit: (code: number | null, signal: string | null) => void;
};

export class NodeRunner {
  readonly name: string;
  readonly use: string;
  private proc: ChildProcess | null = null;
  private options: NodeRunnerOptions;
  private _ready = false;
  private _readyPromiseResolve: (() => void) | null = null;
  private _readyPromise: Promise<void>;

  constructor(options: NodeRunnerOptions) {
    this.name = options.name;
    this.use = options.use;
    this.options = options;
    this._readyPromise = new Promise((resolve) => {
      this._readyPromiseResolve = resolve;
    });
  }

  get ready(): boolean {
    return this._ready;
  }

  /** Wait for this node to emit lifecycle.ready. */
  waitReady(timeoutMs = 10000): Promise<void> {
    if (this._ready) return Promise.resolve();
    return Promise.race([
      this._readyPromise,
      new Promise<void>((_, reject) =>
        setTimeout(
          () => reject(new Error(`Node '${this.name}' did not become ready within ${timeoutMs}ms`)),
          timeoutMs,
        ),
      ),
    ]);
  }

  /** Spawn the node process. */
  spawn(): void {
    const resolved = resolveNode(this.use);
    const env: Record<string, string | undefined> = { ...process.env, ...this.options.env };
    env.ACPFX_NODE_NAME = this.name;
    if (this.options.settings) {
      env.ACPFX_SETTINGS = JSON.stringify(this.options.settings);
    }

    if (resolved.type === "fork") {
      this.proc = fork(resolved.command, resolved.args, {
        stdio: ["pipe", "pipe", "pipe", "ipc"],
        env,
        silent: true,
      });
    } else {
      this.proc = cpSpawn(resolved.command, resolved.args, {
        stdio: ["pipe", "pipe", "pipe"],
        env,
      });
    }

    // Suppress EPIPE errors on stdin (node may exit while we're writing)
    this.proc.stdin!.on("error", () => {});

    // Read NDJSON from stdout
    const rl = createInterface({ input: this.proc.stdout! });
    rl.on("line", (line) => {
      if (!line.trim()) return;
      try {
        const event = parseEvent(line);
        // Check for lifecycle.ready
        if (event.type === "lifecycle.ready") {
          this._ready = true;
          this._readyPromiseResolve?.();
        }
        this.options.onEvent(event);
      } catch (err) {
        this.options.onError(
          new Error(`Node '${this.name}' emitted invalid JSON: ${line}`),
        );
      }
    });

    // Forward stderr
    if (this.use.includes("ui-cli") || this.use.includes("ui-web")) {
      // UI nodes: pipe raw bytes so Ink's ANSI escape sequences work
      this.proc.stderr!.pipe(process.stderr);
    } else {
      // Convert node stderr lines to log events routed through the orchestrator
      const stderrRl = createInterface({ input: this.proc.stderr! });
      stderrRl.on("line", (line) => {
        if (line.includes("buffer underflow") || line.includes("Didn't have any audio")) return;
        if (!line.trim()) return;
        // Emit as a log event so it flows through the orchestrator like any other event
        this.options.onEvent({
          type: "log",
          level: line.toLowerCase().includes("error") ? "error" : "info",
          component: this.name,
          message: line,
        } as AnyEvent);
      });
    }

    this.proc.on("error", (err) => {
      this.options.onError(err);
    });

    this.proc.on("exit", (code, signal) => {
      this.proc = null;
      this.options.onExit(code, signal);
    });
  }

  /** Send an event to this node's stdin. */
  send(event: AnyEvent): void {
    if (!this.proc?.stdin?.writable) {
      if (event.type === "audio.chunk") {
        // Only log once per dropped burst
        if (!this._loggedDrop) {
          this._loggedDrop = true;
          process.stderr.write(`[orchestrator] WARNING: dropping ${event.type} to ${this.name} (stdin not writable)\n`);
        }
      }
      return;
    }
    this._loggedDrop = false;
    try {
      this.proc.stdin.write(serializeEvent(event) + "\n");
    } catch {
      // Node may have exited — ignore write errors
    }
  }
  private _loggedDrop = false;

  /** Gracefully shut down: close stdin, then SIGTERM after timeout. */
  async stop(timeoutMs = 3000): Promise<void> {
    if (!this.proc) return;
    const proc = this.proc;

    // Close stdin to signal EOF
    proc.stdin?.end();

    // Wait for exit, or SIGTERM
    await new Promise<void>((resolve) => {
      const timer = setTimeout(() => {
        proc.kill("SIGTERM");
        resolve();
      }, timeoutMs);
      proc.on("exit", () => {
        clearTimeout(timer);
        resolve();
      });
    });
  }

  /** Forcefully kill the node process. */
  kill(): void {
    this.proc?.kill("SIGKILL");
  }
}

export type ResolvedNode = {
  command: string;
  args: string[];
  type: "fork" | "spawn";
};

/**
 * Resolve a `use` string to a command + launch strategy.
 *
 * For @acpfx/<name>:
 *   1. Local JS:     dist/nodes/<name>.js  → fork()
 *   2. Local binary: dist/nodes/<name>     → spawn()
 *   3. npx fallback: npx -y @acpfx/<name> → spawn()
 *
 * For external paths:
 *   - .js/.mjs → fork()
 *   - otherwise → spawn()
 */
export function resolveNode(use: string): ResolvedNode {
  const match = use.match(/^@acpfx\/(.+)$/);
  if (match) {
    const name = match[1];
    // 1. Local JS bundle
    const jsPath = resolve(__dirname, "nodes", `${name}.js`);
    if (existsSync(jsPath)) {
      return { command: jsPath, args: [], type: "fork" };
    }
    // 2. Local native binary
    const binPath = resolve(__dirname, "nodes", name);
    if (existsSync(binPath)) {
      return { command: binPath, args: [], type: "spawn" };
    }
    // 3. npx fallback
    return { command: "npx", args: ["-y", use], type: "spawn" };
  }

  // External path
  if (use.endsWith(".js") || use.endsWith(".mjs")) {
    return { command: resolve(use), args: [], type: "fork" };
  }
  return { command: resolve(use), args: [], type: "spawn" };
}

File: packages/orchestrator/src/node_runner.rs
//! Node Runner — spawns a node as a child process, manages NDJSON stdin/stdout.
//!
//! Each node is a child process that:
//! - Receives NDJSON events on stdin
//! - Emits NDJSON events on stdout
//! - Logs to stderr (forwarded to parent's stderr)
//! - Gets settings via ACPFX_SETTINGS env var
//! - Must emit lifecycle.ready when initialized
//! - Exits on stdin EOF or SIGTERM

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use std::process::Stdio;

use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::process::{Child, Command};
use tokio::sync::{mpsc, oneshot};

/// How to launch a resolved node.
#[derive(Debug, Clone)]
#[allow(dead_code)]
pub enum LaunchType {
    /// Fork via `node <path>` (JS files)
    Fork,
    /// Spawn directly (native binaries, npx)
    Spawn,
}

/// A resolved node ready to spawn.
#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct ResolvedNode {
    pub command: String,
    pub args: Vec<String>,
    pub launch_type: LaunchType,
}

/// Resolve a `use` string to a command + launch strategy.
///
/// When `dist_dir` is Some (debug/dev), tries local resolution first:
///   1. Local JS:     dist/nodes/<name>.js  -> fork via `node`
///   2. Local binary: dist/nodes/<name>     -> spawn directly
///   3. npx fallback
///
/// When `dist_dir` is None (release), goes straight to npx for @acpfx/* packages.
///
/// External paths always resolve directly regardless of mode.
pub fn resolve_node(use_: &str, dist_dir: Option<&Path>) -> ResolvedNode {
    if let Some(name) = use_.strip_prefix("@acpfx/") {
        // Debug/dev: try local dist/ first
        if let Some(dist) = dist_dir {
            let nodes_dir = dist.join("nodes");

            // 1. Local JS bundle
            let js_path = nodes_dir.join(format!("{name}.js"));
            if js_path.exists() {
                return ResolvedNode {
                    command: "node".into(),
                    args: vec![js_path.to_string_lossy().into()],
                    launch_type: LaunchType::Fork,
                };
            }

            // 2. Local native binary
            let bin_path = nodes_dir.join(name);
            if bin_path.exists() {
                return ResolvedNode {
                    command: bin_path.to_string_lossy().into(),
                    args: vec![],
                    launch_type: LaunchType::Spawn,
                };
            }
        }

        // Release (or local not found): use npx
        return ResolvedNode {
            command: "npx".into(),
            args: vec!["-y".into(), use_.into()],
            launch_type: LaunchType::Spawn,
        };
    }

    // External path
    let path = PathBuf::from(use_);
    if use_.ends_with(".js") || use_.ends_with(".mjs") {
        ResolvedNode {
            command: "node".into(),
            args: vec![
                std::fs::canonicalize(&path)
                    .unwrap_or(path)
                    .to_string_lossy()
                    .into(),
            ],
            launch_type: LaunchType::Fork,
        }
    } else {
        let abs = std::fs::canonicalize(&path).unwrap_or(path);
        ResolvedNode {
            command: abs.to_string_lossy().into(),
            args: vec![],
            launch_type: LaunchType::Spawn,
        }
    }
}

/// Events emitted by a NodeRunner to the orchestrator.
#[derive(Debug)]
#[allow(dead_code)]
pub enum NodeEvent {
    /// A parsed NDJSON event from the node's stdout.
    Event(serde_json::Value),
    /// The node process exited.
    Exited { code: Option<i32>, name: String },
    /// An error occurred.
    Error { message: String, name: String },
}

/// Handle to a running node process.
pub struct NodeRunner {
    name: String,
    /// Channel to send NDJSON events to the node's stdin.
    stdin_tx: mpsc::Sender<String>,
    /// Handle to the spawned child (for shutdown).
    child: Option<Child>,
    ready_rx: Option<oneshot::Receiver<()>>,
}

impl NodeRunner {
    /// Spawn a node process and start reading its stdout/stderr.
    pub fn spawn(
        name: &str,
        resolved: &ResolvedNode,
        settings: Option<&serde_json::Value>,
        env: &BTreeMap<String, String>,
        event_tx: mpsc::Sender<NodeEvent>,
    ) -> Self {
        let mut cmd = Command::new(&resolved.command);
        cmd.args(&resolved.args)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());

        // Set environment
        cmd.env("ACPFX_NODE_NAME", name);
        if let Some(s) = settings {
            cmd.env("ACPFX_SETTINGS", serde_json::to_string(s).unwrap_or_default());
        }
        for (k, v) in env {
            cmd.env(k, v);
        }

        let mut child = cmd.spawn().unwrap_or_else(|e| {
            panic!("Failed to spawn node '{name}' ({:?}): {e}", resolved.command);
        });

        let stdin = child.stdin.take().expect("stdin must be piped");
        let stdout = child.stdout.take().expect("stdout must be piped");
        let stderr = child.stderr.take().expect("stderr must be piped");

        // Stdin writer task
        let (stdin_tx, mut stdin_rx) = mpsc::channel::<String>(256);
        tokio::spawn(async move {
            let mut stdin = stdin;
            while let Some(line) = stdin_rx.recv().await {
                if stdin.write_all(line.as_bytes()).await.is_err() {
                    break;
                }
                if stdin.write_all(b"\n").await.is_err() {
                    break;
                }
                let _ = stdin.flush().await;
            }
        });

        // Ready signal
        let (ready_tx, ready_rx) = oneshot::channel::<()>();
        let mut ready_tx = Some(ready_tx);

        // Stdout reader task — parse NDJSON, detect lifecycle.ready
        let node_name = name.to_string();
        let event_tx_stdout = event_tx.clone();
        tokio::spawn(async move {
            let reader = BufReader::new(stdout);
            let mut lines = reader.lines();
            while let Ok(Some(line)) = lines.next_line().await {
                if line.trim().is_empty() {
                    continue;
                }
                match serde_json::from_str::<serde_json::Value>(&line) {
                    Ok(event) => {
                        // Detect lifecycle.ready
                        if event.get("type").and_then(|t| t.as_str()) == Some("lifecycle.ready") {
                            if let Some(tx) = ready_tx.take() {
                                let _ = tx.send(());
                            }
                        }
                        let _ = event_tx_stdout
                            .send(NodeEvent::Event(event))
                            .await;
                    }
                    Err(_) => {
                        let _ = event_tx_stdout
                            .send(NodeEvent::Error {
                                message: format!("Invalid JSON from stdout: {line}"),
                                name: node_name.clone(),
                            })
                            .await;
                    }
                }
            }
        });

        // Stderr reader task — forward to parent stderr
        let node_name_err = name.to_string();
        tokio::spawn(async move {
            let reader = BufReader::new(stderr);
            let mut lines = reader.lines();
            while let Ok(Some(line)) = lines.next_line().await {
                eprintln!("[{node_name_err}] {line}");
            }
        });

        NodeRunner {
            name: name.to_string(),
            stdin_tx,
            child: Some(child),
            ready_rx: Some(ready_rx),
        }
    }

    /// Wait for this node to emit lifecycle.ready.
    pub async fn wait_ready(&mut self, timeout_ms: u64) -> Result<(), String> {
        if let Some(rx) = self.ready_rx.take() {
            match tokio::time::timeout(
                std::time::Duration::from_millis(timeout_ms),
                rx,
            )
            .await
            {
                Ok(Ok(())) => Ok(()),
                Ok(Err(_)) => Err(format!(
                    "Node '{}' stdout closed before lifecycle.ready",
                    self.name
                )),
                Err(_) => Err(format!(
                    "Node '{}' did not become ready within {timeout_ms}ms",
                    self.name
                )),
            }
        } else {
            Ok(()) // Already received ready
        }
    }

    /// Send an NDJSON event to this node's stdin.
    pub async fn send(&self, json: &str) {
        let _ = self.stdin_tx.send(json.to_string()).await;
    }

    /// Gracefully shut down: close stdin, then SIGTERM after timeout.
    pub async fn stop(&mut self, timeout_ms: u64) {
        // Drop stdin sender to close the pipe
        // (The channel will be closed when NodeRunner is dropped or we replace it)

        if let Some(mut child) = self.child.take() {
            // Close stdin by dropping the sender
            drop(std::mem::replace(
                &mut self.stdin_tx,
                mpsc::channel(1).0,
            ));

            // Wait for exit or kill after timeout
            match tokio::time::timeout(
                std::time::Duration::from_millis(timeout_ms),
                child.wait(),
            )
            .await
            {
                Ok(_) => {} // Exited cleanly
                Err(_) => {
                    // Timeout — kill
                    let _ = child.kill().await;
                }
            }
        }
    }

    #[allow(dead_code)]
    pub fn name(&self) -> &str {
        &self.name
    }
}
